core/server/services/members/importer/index.js

```diff
@@ -80,150 +179,92 @@ const doImport = async ({members, allLabelModels, importSetLabels, imported, inv
+        labelAssociationsToInsert.push(...memberLabelAssociationsToInsert);
 
         if (member.stripe_customer_id) {
-            membersWithStripeCustomers.push({
-                stripe_customer_id: member.stripe_customer_id,
-                id: memberId,
-                email: member.email
-            });
-        }
-
-        if ((String(member.complimentary_plan).toLocaleLowerCase() === 'true')) {
-            membersWithComplimentaryPlans.push({
-                id: memberId,
-                email: member.email
-            });
+            const stripeCustomerToInsert = {
+                customer_id: member.stripe_customer_id,
+                member_id: memberToInsert.id
+            };
+            stripeCustomersToInsert.push(stripeCustomerToInsert);
         }
     });
 
-    try {
-        // TODO: below inserts most likely need to be wrapped into transaction
-        //       to avoid creating orphaned member_labels connections
-        const CHUNK_SIZE = 5000;
-        const chunkedMembers = _.chunk(mappedMemberBatchData, CHUNK_SIZE);
-        for (const data of chunkedMembers) {
-            await db.knex('members')
-                .insert(data);
-        }
-
-        const chunkedLebelAssociations = _.chunk(mappedMembersLabelsBatchAssociations, CHUNK_SIZE);
-
-        for (const data of chunkedLebelAssociations) {
-            await db.knex('members_labels')
-                .insert(data);
-        }
-
-        imported.count += mappedMemberBatchData.length;
-    } catch (error) {
-        logging.error(error);
-
-        if (error.code && error.message.toLowerCase().indexOf('unique') !== -1) {
-            invalid.errors.push(new errors.ValidationError({
-                message: i18n.t('errors.api.members.memberAlreadyExists.message'),
-                context: i18n.t('errors.api.members.memberAlreadyExists.context')
-            }));
-        } else {
-            // NOTE: probably need to wrap this error into something more specific e.g. ImportError
-            invalid.errors.push(error);
-        }
-
-        invalid.count += mappedMemberBatchData.length;
-    }
+    return {
+        invalidMembers,
+        membersToInsert,
+        stripeCustomersToInsert,
+        labelAssociationsToInsert
+    };
+}
 
-    if (membersWithStripeCustomers.length || membersWithComplimentaryPlans.length) {
-        const deleteMemberKnex = async (id) => {
-            // TODO: cascading wont work on SQLite needs 2 separate deletes
-            //       for members_labels and members wrapped into a transaction
-            const deletedMembersCount = await db.knex('members')
-                .where('id', id)
-                .del();
-
-            if (deletedMembersCount) {
-                imported.count -= deletedMembersCount;
-                invalid.count += deletedMembersCount;
-            }
-        };
+async function fetchStripeCustomers(stripeCustomersToInsert) {
+    const LeakyBucket = require('leaky-bucket');
```

Comment:
Awesome one. Definitely a thing to rely upon stead of testing at which concurrency count things break :+1:

Reply:
Yeah! The implementation is pretty big though, has a lot of unnecessary deps :confused: The newest version looks cool, but needs to be run with an experimental flag :weary:  Might be worth writing our own, or finding a better library in the future.